File: Telephone/AKKeychain.m
//
//  AKKeychain.m
//  Telephone
//
//  Copyright © 2008-2016 Alexey Kuznetsov
//  Copyright © 2016-2022 64 Characters
//  Modifications © 2026 Pavel Gordeev
//
//  Telephone is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//

#import "AKKeychain.h"

@import Security;

@implementation AKKeychain

+ (nonnull NSString *)passwordForService:(nonnull NSString *)service account:(nonnull NSString *)account {
    NSParameterAssert(service);
    NSParameterAssert(account);

    NSDictionary *query = @{
        (__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecAttrService: service,
        (__bridge id)kSecAttrAccount: account,
        (__bridge id)kSecMatchLimit: (__bridge id)kSecMatchLimitOne,
        (__bridge id)kSecReturnData: @YES
    };

    CFTypeRef result = NULL;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &result);
    if (status != errSecSuccess || result == NULL) {
        if (result != NULL) {
            CFRelease(result);
        }
        return @"";
    }

    NSData *passwordData = CFBridgingRelease(result);
    NSString *password = [[NSString alloc] initWithData:passwordData encoding:NSUTF8StringEncoding];
    return password ?: @"";
}

+ (BOOL)addItemWithService:(nonnull NSString *)service
                   account:(nonnull NSString *)account
                  password:(nonnull NSString *)password {
    NSParameterAssert(service);
    NSParameterAssert(account);
    NSParameterAssert(password);

    NSData *passwordData = [password dataUsingEncoding:NSUTF8StringEncoding];
    NSDictionary *query = @{
        (__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecAttrService: service,
        (__bridge id)kSecAttrAccount: account
    };

    NSDictionary *attributes = @{
        (__bridge id)kSecValueData: passwordData
    };

    OSStatus updateStatus = SecItemUpdate(
        (__bridge CFDictionaryRef)query,
        (__bridge CFDictionaryRef)attributes
    );
    if (updateStatus == errSecSuccess) {
        return YES;
    }
    if (updateStatus != errSecItemNotFound) {
        return NO;
    }

    NSMutableDictionary *newItem = [query mutableCopy];
    [newItem addEntriesFromDictionary:attributes];
    return SecItemAdd((__bridge CFDictionaryRef)newItem, NULL) == errSecSuccess;
}

@end

File: Telephone/AKKeychain.swift
//
//  AKKeychain.swift
//  Telephone
//
//  Swift implementation of the Objective-C AKKeychain interface.
//

import Foundation
import Security

@objc @implementation
extension AKKeychain {
    static func password(
        forService service: String,
        account: String
    ) -> String {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
            kSecMatchLimit: kSecMatchLimitOne,
            kSecReturnData: true,
        ]

        var result: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        guard
            status == errSecSuccess,
            let data = result as? Data
        else {
            return ""
        }

        return String(data: data, encoding: .utf8) ?? ""
    }

    static func addItem(
        withService service: String,
        account: String,
        password: String
    ) -> Bool {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
        ]
        let attributes: [CFString: Any] = [
            kSecValueData: Data(password.utf8),
        ]

        let updateStatus = SecItemUpdate(
            query as CFDictionary,
            attributes as CFDictionary
        )
        if updateStatus == errSecSuccess {
            return true
        }
        guard updateStatus == errSecItemNotFound else {
            return false
        }

        var newItem = query
        newItem[kSecValueData] = Data(password.utf8)
        return SecItemAdd(newItem as CFDictionary, nil) == errSecSuccess
    }

    static func removeItem(
        forService service: String,
        account: String
    ) -> Bool {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
        ]

        let status = SecItemDelete(query as CFDictionary)
        return status == errSecSuccess || status == errSecItemNotFound
    }
}
